### PR TITLE
i#2390: Replace add+ldar with ldr+dmb in AArch64 HT lookup

### PR DESCRIPTION
**Pull request by merckhung**
Replace the add+ldar with a ldr+dmb pair in HT lookup and results in
-3% to -28% (reduction) in overhead (instrument vs. native) ratios
of SPECInt 2006 and 2017. Although, found a regression in 657.xz_s
model (+7%).

An add+ldar pair was used to prevent memory-access instructions from
being reordered to ensure a hash mask is always loaded before a
hash_table is loaded. The same ordering restriction is also imposed
in the corresponding update_lookuptable_tls() routine.

With the ldr+dmb pair replacement, an add instruction is eliminated
from the critical path and contributed majorly to overhead reductions.

In addition, a frequently taken branch of the inner-loop epilogue is
converged into a more condensed (4 instr.) and smaller inner-loop,
with the cost of adding 1 more sub instruction in both 2 exit paths.

Since exit paths are not as hot as HT lookup, the impact of adding
a sub instruction in both is quite trivial (not seen).

In order to condense the inner-loop down to 4 instructions, a
pre-index ldr is replaced with a post-index version in the prologue.

On the hit-exit path, register use is carefully swapped and results in a
move instruction elimination. The number of instructions on the path
remains unchanged since a sub instruction is added to its entry.

Issue: #2390
Tests: SPECInt 2006 and 2017 ran on ARM Juno r2 (w/ 8GB RAM, Debian)

### CODE REVIEW

File: core/arch/aarch64/emit_utils.c [hunk 1]
```diff
@@ -717,29 +716,33 @@ emit_indirect_branch_lookup(dcontext_t *dc, generated_code_t *code, byte *pc,
...
+    APP(&ilist,
+        INSTR_CREATE_and(dc, opnd_create_reg(DR_REG_X1), opnd_create_reg(DR_REG_X1),
+                         opnd_create_reg(DR_REG_X2)));
+    /* dmb ishld to prevent memory-access instructions from being reordered across the
+     * barrier. DMB results generally in better performances compared to LDAR in the case.
+     */
     APP(&ilist,
-        INSTR_CREATE_ldar(dc, opnd_create_reg(DR_REG_X1),
-                          OPND_CREATE_MEMPTR(DR_REG_X1, 0)));
+        instr_create_0dst_1src(dc, OP_dmb, OPND_CREATE_INT(/* DR_DMB_ISHLD */ 9)));
```
Comment (derekbruening) [inline core/arch/aarch64/emit_utils.c:739]: We'd prefer to use INSTR_CREATE_macros: ARM and x86 are complete but it seems AArch64 is missing some.  Could you add the ones you need?  IIRC they are manually done instead of automated like the ARM ones are.

Comment (derekbruening) [inline core/arch/aarch64/emit_utils.c:739]: Could we use `DR_DMB_ISHLD` directly?

File: core/arch/aarch64/emit_utils.c [hunk 2]
```diff
@@ -750,11 +753,13 @@ emit_indirect_branch_lookup(dcontext_t *dc, generated_code_t *code, byte *pc,
     /* x1 now holds the fragment_entry_t* in the hashtable. */
     APP(&ilist, load_tag);
     /* Load tag from fragment_entry_t* in the hashtable to x0. */
-    /* ldr x0, [x1, #tag_fragment_offset] */
-    APP(&ilist,
-        INSTR_CREATE_ldr(
-            dc, opnd_create_reg(DR_REG_X0),
-            OPND_CREATE_MEMPTR(DR_REG_X1, offsetof(fragment_entry_t, tag_fragment))));
+    /* ldr x0, [x1], #tag_fragment_offset */
+    instr_t *instr = instr_create_2dst_3src(
```
Comment (derekbruening) [inline core/arch/aarch64/emit_utils.c:757]: We'd again prefer an INSTR_CREATE_.

File: core/arch/aarch64/emit_utils.c [hunk 3]
```diff
@@ -717,29 +716,33 @@ emit_indirect_branch_lookup(dcontext_t *dc, generated_code_t *code, byte *pc,
...
      */
-
     /* Spill x0. */
-    APP(&ilist, instr_create_save_to_tls(dc, DR_REG_R0, TLS_REG3_SLOT));
-    /* Load-acquire hash mask.  We need a load-acquire to ensure we see updates
-     * properly; the corresponding store-release is in update_lookuptable_tls().
+    APP(&ilist, instr_create_save_to_tls(dc, DR_REG_X0, TLS_REG3_SLOT));
+    /* Load hash mask before loading hash_table, with DMB to prevent memory-access
+     * instructions from being reordered. A corresponding store-release is in
+     * update_lookuptable_tls() to ensure a new hast_table will be written after a
```
Comment (derekbruening) [inline core/arch/aarch64/emit_utils.c:723]: typo: hash_table

**Comment by derekbruening** [comment]
> I took a quick look. Probably best for an AArch64 expert and/or those who wrote the original code to take a deeper look.

Requesting @egrimley to take a look (or delegate)

**Comment by AssadHashmi** [comment]
run arm tests

**Comment by AssadHashmi** [comment]
run arm tests

File: core/arch/aarch64/emit_utils.c (hunk 1 above)
Comment (merckhung) [reply core/arch/aarch64/emit_utils.c]: Got it. I have created INSTR_CREATE_dmb macro for AArch64 in the 2nd commit.

Comment (merckhung) [reply core/arch/aarch64/emit_utils.c]: Sure thing. I have made a copy in AArch64's instr_create.h instead of including ARM's header, to avoid macro definition conflicts across architectures.

File: core/arch/aarch64/emit_utils.c (hunk 2 above)
Comment (merckhung) [reply core/arch/aarch64/emit_utils.c]: I see. INSTR_CREATE_ldr_imm has been created to serve this purpose.

File: core/arch/aarch64/emit_utils.c (hunk 3 above)
Comment (merckhung) [reply core/arch/aarch64/emit_utils.c]: Corrected. Thank you!

**Comment by derekbruening** [comment]
@AssadHashmi -- trying to get a review from an ARM expert

**Comment by AssadHashmi** [comment]
> @AssadHashmi -- trying to get a review from an ARM expert

Based on my (limited) experience the change looks functionally correct.I can't comment on performance

File: core/ir/aarch64/instr_create.h [hunk 4]
```diff
@@ -2514,5 +2520,28 @@
...
     instr_create_1dst_3src(dc, OP_umull2, Rd, Rm, Rn, width)
 
+/**
+ * Creates a DMB instruction.
+ * \param dc      The void * dcontext used to allocate memory for the instr_t.
+ * \param imm     The integer constant opnd_t operand.
+ */
+#define INSTR_CREATE_dmb(dc, imm) instr_create_0dst_1src((dc), OP_dmb, (imm))
+
+/** Immediate values for INSTR_CREATE_dmb(). */
```
Comment (derekbruening) [inline core/ir/aarch64/instr_create.h:2530]: This looks like an exact copy of the enum in arm/.  To avoid duplicating code, please instead move it to a shared header.  Maybe instr_create_shared.h, or maybe a new ir/aarchxx/instr_create_aarchxx.h, or possibly instr.h or opnd.h.

File: core/ir/aarch64/instr_create.h [hunk 5]
```diff
@@ -2514,5 +2520,28 @@
 #define INSTR_CREATE_umull2_vector(dc, Rd, Rm, Rn, width) \
     instr_create_1dst_3src(dc, OP_umull2, Rd, Rm, Rn, width)
 
+/**
+ * Creates a DMB instruction.
+ * \param dc      The void * dcontext used to allocate memory for the instr_t.
+ * \param imm     The integer constant opnd_t operand.
```
Comment (derekbruening) [inline core/ir/aarch64/instr_create.h:2526]: Point at one of the DR_DMB_OSHLD values, using `#` to make a doxygen link, so the user knows there are named constants.

File: core/ir/opnd.h [hunk 6]
```diff
@@ -1702,7 +1702,7 @@ DR_API
 opnd_t
 opnd_create_far_base_disp_ex(reg_id_t seg, reg_id_t base_reg, reg_id_t index_reg,
                              int scale, int disp, opnd_size_t size, bool encode_zero_disp,
-                             bool force_full_disp, bool disp_short_addr);
+                             bool force_full_disp, bool disp_short_addr, bool post_index);
```
Comment (derekbruening) [inline core/ir/opnd.h:1705]: This is a breaking interface change.  This needs to be accompanied by documentation changes announcing the break.  Alternatively, instead of changing this function, a separate opnd_set_zero_offset_pre_index() could be added, which would be better for compatibility.

Comment (derekbruening) [reply core/ir/opnd.h:1705]: I guess opnd_set_zero_offset_post_index() since the default is pre.

File: core/ir/aarch64/instr_create.h [hunk 7]
```diff
@@ -539,6 +539,12 @@
 #define INSTR_CREATE_ldp(dc, rt1, rt2, mem) \
     instr_create_2dst_1src(dc, OP_ldp, rt1, rt2, mem)
 #define INSTR_CREATE_ldr(dc, Rd, mem) instr_create_1dst_1src((dc), OP_ldr, (Rd), (mem))
+#define INSTR_CREATE_ldr_imm(dc, Rt, Rn, imm, post)                                      \
```
Comment (derekbruening) [inline core/ir/aarch64/instr_create.h:542]: Not sure why this needs the "post": that is just an encoding detail that doesn't affect semantics, and it only applies to a zero offset, right?

File: core/ir/aarch64/instr_create.h (hunk 4 above)
Comment (derekbruening) [reply core/ir/aarch64/instr_create.h:2530]: (Wondering what happens to the duplicated enum values in the Doxygen generated docs?  Is there a warning from Doxygen?)

File: core/ir/aarch64/instr_create.h (hunk 7 above)
Comment (derekbruening) [reply core/ir/aarch64/instr_create.h:542]: My comment is based on only passing this to opnd_create_far_base_disp_ex().  See the comment for the pre_index field in opnd.h.  For creating semantically-different variants of the instruction itself, those are going to change the other operands too and not just the memop, right?  Not sure what is currently done for other A64 instrs, but for A32 we have different variants of the macro itself for different write-back options.

File: core/ir/aarch64/instr_create.h (hunk 4 above)
Comment (merckhung) [reply core/ir/aarch64/instr_create.h:2530]: Got it. I will change the code accordingly. It will take a while on my end, but I will submit a version by the end of Wed. I didn't look at the generated Doxygen docs, let me check it out. I don't remember it warns, perhaps I didn't notice one back then.

Comment (merckhung) [reply core/ir/aarch64/instr_create.h:2530]: I have had a look of those aforementioned files. Thank you!
instr.h: defines architectural values WRT instruction encoding/decoding.
opnd.h: defines architectural values WRT operand and register encoding/decoding.
instr_create_shared.h: seems like defining architecturally independent macros and definitions.

For the DMB macro and its IMM definitions, I feel creating a new ir/aarchxx/instr_create_aarchxx.h is a way to go.

Because DMB's IMM enum doesn't share the exactly same semantics with the other arch (e.g. x86), so the instr_create_shared.h is ruled out;

instr.h and opnd.h are more related to architectural definitions, rather than instruction-level definitions. So I think I will go with creating a new ir/aarchxx/instr_create_aarchxx.h and move both ARM and AArch64's DMB macro and IMM enum to the new file. If no objection, then I will go with this path forward. Thank you.

File: core/ir/aarch64/instr_create.h (hunk 7 above)
Comment (merckhung) [reply core/ir/aarch64/instr_create.h:542]: I should've read the explanation of the "pre_index" bitfield. Thank you.
I did gdb tracing and found pre_index is hard-coded to true, so I figured the way to workaround it is to post-process it (manually changing to false, after encoded).

Today, I've tried to passing a negative offset to opnd_create_far_base_disp_ex() and I found it doesn't encode correctly.
Report: Internal Error: DynamoRIO debug check failure: /home/merck/Code/dynamorio_orig/core/arch/emit_utils_shared.c:1061 nxt_pc != NULL

** I will look into it tomorrow to see what actually happens, but I guess the AArch64 encoder doesn't catch it.

I realized that when a pre_index LDR (immediate) is encoded the instr_t will be handled by encode_opndsgen_38400400() in encode_gen.h (line: 8085). The AArch64 encoder tries each possible encoding functions until one of those catches it.

LDR (immediate) differs pre- or post-index by looking at bit 11 and 12.
When it comes to encoding a post-index instruction, the encoder will fall through to next encode_opndsgen_38400c00() function (bit11 = 1), and the instr_t is supposed to be caught by it.
In the first commit of this pull request, I found toggling the pre_index bitfield does affect how AArch64 encodes LDR (immediate). But passing a negative offset to opnd_create_far_base_disp_ex() doesn't seem be encoded correctly in AArch64.

Let me have a look tmr, and I will come back with a better change for reviews. Thank you!

Comment (merckhung) [reply core/ir/aarch64/instr_create.h:542]: BTW, thank you for laying out the design principles.

Comment (derekbruening) [reply core/ir/aarch64/instr_create.h:542]: > Today, I've tried passing a negative offset to opnd_create_far_base_disp_ex() and I found it doesn't encode correctly.
> Report: Internal Error: DynamoRIO debug check failure: /home/merck/Code/dynamorio_orig/core/arch/emit_utils_shared.c:1061 nxt_pc != NULL

I remember something off in the A64 branch linking code: #4273, but that's a different assert.  @AssadHashmi might know more?

File: core/ir/aarch64/instr_create.h (hunk 4 above)
Comment (derekbruening) [reply core/ir/aarch64/instr_create.h:2530]: ir/aarchxx/instr_create_aarchxx.h seems good to me.

File: core/ir/aarch64/instr_create.h (hunk 7 above)
Comment (merckhung) [reply core/ir/aarch64/instr_create.h:542]: I've done some investigations about the post-LDR encoding problem.

Summary:
0) I think the explanation of pre_index bit-field in the opnd.h is described in terms of A64 decoding.
   For recording a decoded LDR is pre- or post-index, it uses the pre_index bit-field to distinguish it.
   The value of the pre_index will be written by A64 decoder according to the input app's code sequence,
   and the JIT is supposed to emit a correct LDR version by looking at the value of the pre_index field.
   
   It's not meant to say a post-index LDR cannot have an offset other than 0, but both pre- and post-index
   LDR could both have a zero offset encoded. So the pre_index field is needed for distinguishing them in DR.

   Therefore, for *manually* encoding an LDR in the DR program itself, the pre_index field is hard-coded to *true*.
   So if a developer calls OPND_CREATE_MEMPTR()->opnd_create_far_base_disp_ex() in DR.
   The JIT will never have a chance to emit a correct post-LDR instr.

Tl;Dr
1) Only LDR/STR family uses encode_opnd_mem9/encode_opnd_mem9post/encode_opnd_mem9_bytes functions to encode the signed offset imm9 field (bit 20:12). If a change happens in those will only affect LDR/STR family.

2) In the encode_opnd_mem9post() and the following encode_opnd_mem9_bytes() function, if post==true, then the disp is requested to be 0, otherwise it will fail the encoding (codec.c, line: 662). *This is where it fails.*
    if (post ? (disp != 0) : (disp < -256 || disp > 255))
        return false;

3) Based on point #1, regardless of the indexing direction, per A64 SPEC. both pre- and post-index LDR/STR support encoding a signed 0 imm9 offset (AARM, p. 1000). But in the point #2, current impl. limits an input of post==true to have a zero disp and allows post==false to have a disp of range from -256 to +255. *This impl. is contradicting with the A64 SPEC.*

4) Because the imm9 field's type is a signed integer. If passing an negative offset to opnd_create_far_base_disp_ex() function for the purpose of determining the pre- or post-index attribute of an LDR/STR encoding will result in prohibiting negative offsets to be correctly encoded in the further pre-index LDR/STR.
*Therefore, will not go this way, but will think about setting FLAG_NEGATIVE and converting value to positive*

5) The later AArch64 instruction encoding process doesn't determine a pre- or post-index LDR by directly looking at the pre_index bit-field. Instead, the A64 encoder tries each possible LDR encoding functions until one of those reports successful. *A bit inefficient in A64 encoding process.* Furthermore, if a preceding encoding function returns TRUE, it may cascadingly block a later *more correct* one. This actually complicates the constructions of opnd_t variables, because the relationship among those opnd_t variables determines the selection of the later A64 encoding functions. Rather than simply determining by looking at the values of individually opnd_t vars at first.

Conclusion:
6) I will study more about ARM's implementation (NEGATIVE FLAG, positive conversion), before I implement a change for A64. Will come back with a more considerate commit that is more aligned with current ARM port.

File: core/ir/aarch64/instr_create.h (hunk 4 above)
Comment (merckhung) [reply core/ir/aarch64/instr_create.h]: I have created instr_create_aarchxx.h in the latest commit, but I am struggling with making it appear in the Doxygen htmls.
Do you know how to make instr_create_aarchxx.h show up in the doxygen updates? Thank you.

File: core/ir/aarch64/instr_create.h (hunk 5 above)
Comment (merckhung) [reply core/ir/aarch64/instr_create.h:2523]: Done in the latest commit, but the Doxygen is not generated correctly.

File: core/ir/opnd.h (hunk 6 above)
Comment (merckhung) [reply core/ir/opnd.h]: Done the change accordingly in the latest commit.

File: core/ir/aarch64/instr_create.h (hunk 7 above)
Comment (merckhung) [reply core/ir/aarch64/instr_create.h]: Reverted the changes of opnd_create_far_base_disp_ex(). Refined INSTR_CREATE_ldr_imm() macros to align with DR design (moved away from AArch64 mnemonics style to DR style). If the developers want a post-indexed LDR, they are supposed to call opnd_set_zero_offset_pre_index() on the just generated memop to set the pre_index field from true to false (indicating post-indexed), see an example in core/arch/aarch64/emit_utils.c.

**Comment by merckhung** [comment]
Close this pull request since I messed up with my fork with a force-push. A clean and new pull request has been created at #4390 4390.

File: core/ir/aarch64/instr_create.h (hunk 4 above)
Comment (derekbruening) [reply core/ir/aarch64/instr_create.h]: It needs to be added to the file list in genapi.pl.
